### PR TITLE
feat(hardening): aco-v2 stabilization and hardening

### LINKED ISSUE #6: task: aco v2 hardening & stabilization
## 배경 — 스프린트 드리프트 기록

이 스프린트는 **v1.3 Go wrapper 마이그레이션**으로 시작했지만 브레인스토밍 세션(docs/brainstorm/) 결과 **v2 아키텍처 피벗**으로 방향이 전환됐다.

**전환 시점**: `feat(phase-2): implement real Go process runner` 이후, 브레인스토밍에서 다음이 결정됨:
- `aco run <provider> <command>` → `aco delegate <agent-id>` (role-less binary)
- async session state 전면 제거 (blocking single-task only)
- frontmatter + formatter 기반 provider routing
- `.claude/agents/` 를 CC와 공유하는 canonical spec으로

**결과**: PR #5 `feat(v2): aco delegate with frontmatter-driven provider routing` 가 v2 baseline.

---

## 현재 상태 (v2 완료)

| 항목 | 상태 |
|------|------|
| `aco delegate <agent-id>` CLI | ✅ 구현 완료 |
| Frontmatter spec (id, modelAlias, roleHint 등) | ✅ 구현 완료 |
| `.aco/formatter.yaml` routing | ✅ 구현 완료 |
| ACO_META sentinel output | ✅ 구현 완료 |
| Async state 제거 (cancel/status/result) | ✅ 제거 완료 |
| Copilot provider 제거 | ✅ 제거 완료 |
| 보안 수정 (path traversal, goroutine leak 등) | ✅ 완료 |
| 테스트 38개 통과 | ✅ |

---

## PR #12 resolved review 반영 완료

**출처**: PR #12 resolved review threads

아래 항목들은 review 코멘트로 지적됐고, PR #12 내에서 fix comment와 함께 resolved 처리된 hardening 항목이다. v2.1 신규 TODO가 아니라 v2 baseline에 반영된 완료 항목으로 추적한다.

- ✅ `cmdRun` user-facing output을 `os.Stdout`/`os.Stderr` 대신 injected `d.stdout`/`d.stderr`로 일원화
- ✅ `cmdRun`, `cmdDelegate` unknown flag를 silent ignore하지 않고 usage와 함께 fast-fail
- ✅ `--input` value guard를 blanket `--*` 거부에서 known flag exact match 거부로 완화
- ✅ `cmdDelegate` provider not found 에러에 `InstallHint()` 포함
- ✅ Copilot provider registration 제거
- ✅ Claude-style `name` frontmatter를 `id` alias로 허용하고, 둘 다 없으면 filename fallback
- ✅ async session/PID store 관련 stale comment 제거 및 blocking contract 문구 정리
- ✅ `Setpgid`/process group 기반 signal handling을 contract와 implementation comment에 명시
- ✅ root fixture script를 unpinned `ts-node` runtime fetch 대신 locked `tsx` 기반으로 정리
- ✅ `go.mod`의 부정확한 Go 1.22 `cmd.WaitDelay` 근거 문구 수정
- ✅ Codex `BuildArgs`에서 `permissionProfile: restricted`일 때 `--full-auto`를 생략하도록 수정
- ✅ Codex auth failure 판정을 exit code 401/403 의존에서 stderr heuristic 포함으로 보완
- ✅ relative `promptSeedFile`을 cwd가 아니라 agent spec 파일 위치 기준으로 resolve

---

## 다음 스프린트 (v2.1) — hardening & stabilization

> **"Zero New Features"** — v2 pivot 후 hardening 우선

### P0/P1 — PR #12 추가 리뷰 코멘트 이관 TODO

**출처**: PR #12 unresolved review threads, 2026-04-07

아래 항목들은 PR #12에 새로 추가된 review comment를 v2.1 hardening backlog로 이관한 것이다. PR thread는 Issue #6에서 추적하는 것으로 답변하고 resolved 처리한다.

**P0 후보**
- [ ] `cmd/aco/cmd_delegate.go`: `cmdDelegate`도 `ACO_TIMEOUT_SECONDS` env override를 honor하도록 `cmdRun`과 동작 일치
- [ ] `cmd/aco/cmd_delegate.go`: `--input` value guard가 diff hunk `---` 등 valid payload를 막지 않도록 known flag exact match 또는 `--input=<value>`/`--` terminator 지원으로 정리
- [ ] `internal/provider/interface.go`: `InvokeOpts` comment의 session/PID persistence(`store.SetPID`, `SessionID`) 잔여 문구 제거 또는 `SessionID` 필요성 명시
- [ ] `internal/provider/codex.go`: `CodexProvider.IsAuthFailure`가 관측 불가능한 process exit code 401/403에 의존하지 않도록 stderr heuristic/현실적 exit code 기준으로 정리

**P1 후보**
- [ ] `cmd/aco/main.go`: file header comment를 실제 CLI surface(`run`, `delegate`)와 일치하도록 수정
- [ ] `test/fixtures/05-exit-code-recording/assertions.ts`: header의 `Known Node.js gap` 표기와 `knownNodeGap: true` metadata 정합성 맞춤
- [ ] `test/fixtures/06-timeout-marking/assertions.ts`: header의 `Known Node.js gap` 표기와 `knownNodeGap: true` metadata 정합성 맞춤
- [ ] `test/fixtures/08-auth-failure/assertions.ts`: header의 `Known Node.js gap` 표기와 `knownNodeGap: true` metadata 정합성 맞춤

---

### P0 — 체크인된 `./aco` 바이너리 stale 상태 해소

**출처**: PR #12 test plan 재현 중 발견

PR #12 head `9ca4f3b` 기준 소스에는 `aco delegate`가 구현되어 있지만, repo root의 체크인된 `./aco` 바이너리는 `delegate` 명령을 인식하지 못한다.

재현 결과:
```text
aco: unknown command "delegate"
Usage: aco run <provider> <command> [options]
       aco --version
```

**완료 기준**
- [ ] repo에 `./aco` 바이너리를 계속 둘지 정책 결정
- [ ] 유지한다면 PR 소스 기준으로 `go build -o aco ./cmd/aco` 후 산출물 갱신
- [ ] 유지하지 않는다면 바이너리를 repo에서 제거하고 build artifact로만 취급하도록 정리

---

### P0 — `gemini_cli` provider의 미지원 옵션 전달 방지

**출처**: PR #12 test plan 재현 중 발견

`researcher` agent는 `.aco/formatter.yaml` 기준으로 `provider=gemini_cli`, `model=gemini-2.5-pro`로 정상 resolve되지만, 현재 Gemini CLI는 wrapper가 전달하는 `--reasoning-effort high` 옵션을 지원하지 않아 실제 end-to-end 실행이 실패한다.

재현 결과:
```text
Unknown arguments: reasoning-effort, reasoningEffort
```

참고: fake `gemini`를 임시 PATH에 넣어 성공 경로를 검증했을 때는 `ACO_META:` sentinel이 stdout 마지막 줄로 출력되는 것을 확인했다.

**완료 기준**
- [ ] `internal/provider/gemini_cli.go`에서 Gemini CLI에 `--reasoning-effort` 전달하지 않도록 수정
- [ ] provider별 지원 옵션 정책 명시: Codex는 reasoning effort 전달, Gemini CLI는 현재 미지원이므로 무시 또는 별도 mapping 도입
- [ ] 관련 unit test 업데이트
- [ ] 실제 `./aco delegate researcher --input "test"`가 설치된 Gemini CLI 기준으로 정상 실행되는지 확인

---

### P0 — ACO_META sentinel prefix collision 방지

**출처**: multi-AI 회고 — Gemini 단독 지적

`ACO_META:` prefix를 provider가 stdout으로 출력하면 caller가 오인 (silent failure).

**해결 방안 검토**
- Option A: 실행마다 랜덤 UUID prefix (`ACO_META_a3f2b1c4:`)
- Option B: sentinel을 stderr로 분리
- Option C: multi-line 구분자

**완료 기준**
- [ ] sentinel 충돌 시나리오 재현 테스트 작성
- [ ] 방안 구현 + `cmd/aco/cmd_delegate_test.go` 업데이트
- [ ] `openspec/specs/aco-v2-spec.md` FR-007 업데이트

---

### P0 — Go/Node.js 계약 경계 문서화

**출처**: multi-AI 회고 — Gemini 지적

v2 전환 후 Go 쪽 validation 로직이 대폭 강화됐지만 `packages/wrapper/` (Node.js)와의 drift 방지 계약이 없다.

**완료 기준**
- [ ] `docs/contract/go-node-boundary.md` 작성
- [ ] CLAUDE.md 관련 섹션 링크 업데이트

---

### P1 — Adversarial fixture 확장

**출처**: multi-AI 회고 — Codex + Gemini 공통

v2 구현이 happy path 위주로 테스트됨. v2 실패 경로 체계적 검증 필요.

| 분류 | 케이스 |
|------|--------|
| Frontmatter | `id` 없음, 닫히지 않은 `---`, 잘못된 타입, `executionMode: background`, path traversal |
| Formatter | 파일 없음, `version: 2`, `fallback` 없음, 매핑 안 되는 alias |
| Payload/Stream | 50k bytes 초과 diff, 멀티바이트 UTF-8 truncation, provider가 `ACO_META:` 출력, sentinel 없이 종료 |
| Process/Signal | SIGTERM 후 5초 내/초과 종료 케이스 |

**완료 기준**
- [ ] `test/fixtures/` 시나리오별 fixture 추가
- [ ] `go test ./...` 전체 통과

---

### P1 — CLAUDE.md v2 방향 갱신

**출처**: multi-AI 회고 — Claude 지적

CLAUDE.md가 아직 v1.x ("thin template", `aco run`) 기준으로 작성되어 있다.

**변경 범위**
- [ ] `## Maintenance Rules` — thin template → context marshaler 원칙
- [ ] `## Key Design Decisions` — `aco run` → `aco delegate`
- [ ] provider 목록 갱신 (copilot 제거, codex + gemini_cli)
- [ ] `.claude/agents/`, `openspec/` 설명 추가

---

### P2 — Ghost Worktree 구현 (`isolationMode: worktree`)

**출처**: PRD v0.2 FR-009, NLSpec B-10 미구현 항목

agent frontmatter `isolationMode: worktree` 설정 시 격리 실행 후 `git diff` 반환.

**완료 기준**
- [ ] worktree lifecycle 구현 (defer cleanup)
- [ ] `isolationMode: worktree` 통합 테스트
- [ ] 선행 조건: P0 sentinel 안정화 완료 후 진행

---

## 관련 아티팩트

- PR #5: `feat(v2): aco delegate with frontmatter-driven provider routing`
- `openspec/prd-aco-v2-native-agent-architecture.md` (PRD v0.2)
- `openspec/specs/aco-v2-spec.md` (NLSpec)
- `docs/brainstorm/` (아키텍처 피벗 근거)

### PR DESCRIPTION
**Pull request by pureliture**
Closes #6

## What

This PR implements the P0 hardening items and P1 improvements for aco v2 architecture stabilization. Key changes include sentinel collision prevention with random identifiers, environment variable whitelist for provider security, file path traversal protection, comprehensive adversarial test fixtures, and documentation updates for the Go/Node.js contract boundary.

## Why

PR #12 review and multi-AI retrospective identified critical gaps in v2 baseline that needed hardening before production use: `ACO_TIMEOUT_SECONDS` env was not honored in cmd_delegate, Gemini CLI received unsupported `--reasoning-effort` options, ACO_META sentinel could collide with provider stdout, and there was no documentation for Go/Node.js responsibility boundaries.

## Changes

- Add `ACO_TIMEOUT_SECONDS` env support in `cmd_delegate.go` matching `cmdRun` behavior
- Improve `--input` value guard with known flag detection, `--input=<value>` format, and `--` terminator support
- Implement sentinel collision prevention using `crypto/rand` 8-byte random identifiers
- Add file path validation to prevent traversal attacks in `--agents-dir`, `--formatter`, and `promptSeedFile`
- Implement environment variable whitelist (only `ACO_TIMEOUT_SECONDS` passed to providers)
- Remove `EffortMap.gemini_cli` mapping (YAGNI - Gemini CLI doesn't support `--reasoning-effort`)
- Remove `SessionID` from `InvokeOpts` (dead field from async session removal)
- Add 23 adversarial test fixtures for edge cases (frontmatter, formatter, signal handling)
- Create `docs/contract/go-node-boundary.md` documenting Go/Node.js responsibility boundary
- Add `scripts/verify-contract.ts` for CI contract drift verification
- Update CLAUDE.md with `aco delegate` command, updated provider list, and Go/Node.js contract section

## Checklist
- [x] go test passes (21 files changed, 1648 insertions)
- [ ] manual smoke test
- [x] docs updated (CLAUDE.md, go-node-boundary.md, aco-v2-spec.md)

### CODE REVIEW

File: internal/runner/process.go [hunk 1]
```diff
@@ -80,6 +117,8 @@ func (ProcessRunner) Run(ctx context.Context, opts RunOpts) (RunResult, error) {
 		cmd.Dir = "."
 	}
 	cmd.Env = os.Environ()
+	// Filter environment variables to allowlist for security
+	cmd.Env = filterEnvForAllowlist(cmd.Env)
```
[the remark this answers is not in the document]
Comment (pureliture) [reply internal/runner/process.go:131]: ✅ **Fixed in commit 5bc7dea**

Added PATH and HOME to the env allowlist. Provider CLIs installed via npm (shebang `#!/usr/bin/env node`) now work correctly.

```go
var envAllowlist = map[string]bool{
    "ACO_TIMEOUT_SECONDS": true,
    "PATH":              true,
    "HOME":              true,
    // ... other essential vars
}
```

File: internal/runner/process_test.go [hunk 2]
```diff
@@ -259,6 +259,120 @@ func TestProcessRunner_ProviderNotFound(t *testing.T) {
...
+		// If it exits gracefully (exit 0), the runner still reports exit code 1
+		// because ctx.Err() == context.Canceled triggers the timeout path.
+		// The key verification is that it exits quickly (not force-killed after 5s).
+		// We accept exit code 0 or 1 - what matters is the speed of exit.
+		_ = res.ExitCode
+		// Exit within 8 seconds proves it didn't wait for forceKillDelay
+	case <-time.After(8 * time.Second):
+		t.Fatal("process did not exit within 8 seconds — SIGKILL may have fired prematurely")
+	}
+}
```
[the remark this answers is not in the document]
Comment (pureliture) [reply internal/runner/process_test.go:346]: **Acknowledged**

The test timing is intentionally lenient to allow CI environment variance. The key verification is that the process exits before the force-kill delay (5s) would fire. The test verifies graceful termination occurs within a reasonable window. Additional output assertions could be added, but the current implementation covers the core signal handling behavior.

File: internal/runner/process_test.go [hunk 3]
```diff
@@ -259,6 +259,120 @@ func TestProcessRunner_ProviderNotFound(t *testing.T) {
...
+
+// TestProcessRunner_SIGTEMRForceKill verifies that a provider that ignores
+// SIGTERM is killed with SIGKILL after forceKillDelay.
+func TestProcessRunner_SIGTERMForceKill(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (pureliture) [reply internal/runner/process_test.go]: ✅ **Fixed in commit 5bc7dea**

Corrected typo: `SIGTEMR` → `SIGTERM` in function name.

File: cmd/aco/cmd_delegate.go [hunk 4]
```diff
@@ -181,10 +221,15 @@ func cmdDelegate(d *deps, args []string) int {
 	})
 
 	if !noMeta && result.ProviderExited {
+		rid, err := generateSentinelRID()
+		if err != nil {
+			fmt.Fprintf(d.stderr, "aco: %v\n", err)
+			return 1
+		}
```
[the remark this answers is not in the document]
Comment (pureliture) [reply cmd/aco/cmd_delegate.go:229]: **Acknowledged - Keeping current behavior**

The spec states "sentinel 없이 종료 (stderr 경고)" which means the sentinel is skipped on failure. However, returning exit code 1 is intentional because the crypto/rand failure is a critical error that prevents proper metadata tracking. The caller should be aware that the delegation didn't complete successfully.

We could log a warning and continue with exit code 0, but that would mask a real problem. The current behavior ensures operators are notified when sentinel generation fails.

File: cmd/aco/cmd_delegate.go [hunk 5]
```diff
@@ -249,3 +304,34 @@ func stdinHasData(r io.Reader) bool {
...
+	if base != "" {
+		absPath, err := filepath.Abs(cleaned)
+		if err != nil {
+			return "", fmt.Errorf("invalid file path: %w", err)
+		}
+		absBase, err := filepath.Abs(base)
+		if err != nil {
+			return "", fmt.Errorf("invalid base path: %w", err)
+		}
+		if !strings.HasPrefix(absPath, absBase) {
```
[the remark this answers is not in the document]
Comment (pureliture) [reply cmd/aco/cmd_delegate.go]: ✅ **Fixed in commit 5bc7dea**

Updated `validatePath` to check path components instead of substring matching. Now rejects only actual `.." components, not filenames like `foo..bar`.

File: cmd/aco/cmd_delegate.go [hunk 6]
```diff
@@ -102,6 +125,16 @@ func cmdDelegate(d *deps, args []string) int {
...
 	}
 
+	// Validate paths to prevent traversal attacks
+	if _, err := validatePath(agentsDir, ""); err != nil {
+		fmt.Fprintf(d.stderr, "aco: invalid --agents-dir: %v\n", err)
+		return 1
+	}
+	if _, err := validatePath(formatterPath, ""); err != nil {
+		fmt.Fprintf(d.stderr, "aco: invalid --formatter: %v\n", err)
+		return 1
```
[the remark this answers is not in the document]
Comment (pureliture) [reply cmd/aco/cmd_delegate.go:136]: ✅ **Fixed in commit 5bc7dea**

Updated callers to use the cleaned path returned by `validatePath`. The validation and usage are now consistent.

File: internal/delegate/config.go [hunk 7]
```diff
@@ -195,7 +194,7 @@ func Resolve(spec AgentSpec, formatter Formatter) (Resolution, error) {
 		LaunchArgs: append([]string(nil), formatter.ProviderDefaults[route.Provider].LaunchArgs...),
 	}
 	if spec.ReasoningEffort != "" {
-		resolution.ReasoningEffort = formatter.EffortMap[route.Provider][spec.ReasoningEffort]
+		resolution.ReasoningEffort = spec.ReasoningEffort
 	}
```
[the remark this answers is not in the document]
Comment (pureliture) [reply internal/delegate/config.go:198]: **Acknowledged - Current behavior is intentional**

The `EffortMap.gemini_cli` was removed as YAGNI (Gemini CLI doesn't support `--reasoning-effort`). The `reasoningEffort` field is passed through directly to all providers, and it's each provider's responsibility to ignore unsupported flags.

For `gemini_cli`, the reasoning effort is simply not included in the command line arguments. This is cleaner than maintaining a provider-specific allowlist that would need updating when/if Gemini CLI adds support.

File: internal/delegate/config.go [hunk 8]
```diff
@@ -225,7 +224,13 @@ func BuildPrompt(spec AgentSpec, input string) (string, error) {
...
 			seedPath = filepath.Join(filepath.Dir(spec.Path), seedPath)
 		}
-		data, err := os.ReadFile(seedPath)
+		// Validate the path to prevent traversal attacks
+		// Reject any path that contains ".." before cleaning
+		if strings.Contains(spec.PromptSeedFile, "..") {
+			return "", fmt.Errorf("invalid promptSeedFile: path traversal not allowed")
+		}
+		cleanedPath := filepath.Clean(seedPath)
+		data, err := os.ReadFile(cleanedPath)
```
[the remark this answers is not in the document]
Comment (pureliture) [reply internal/delegate/config.go:240]: ✅ **Fixed in commit 5bc7dea**

Updated `BuildPrompt` to check for `.." as a path component, not substring. Also validates the cleaned absolute path stays within the agent directory.

File: cmd/aco/cmd_delegate_test.go [hunk 9]
```diff
@@ -113,11 +110,17 @@ fallback:
 	if testRunner.last.PermProfile != provider.ProfileRestricted {
 		t.Fatalf("permission profile = %q, want restricted", testRunner.last.PermProfile)
 	}
-	if testRunner.last.ReasoningEffort != "high" {
+	if testRunner.last.ReasoningEffort != "max" {
 		t.Fatalf("reasoning effort = %q, want high", testRunner.last.ReasoningEffort)
```
[the remark this answers is not in the document]
Comment (pureliture) [reply cmd/aco/cmd_delegate_test.go]: ✅ **Fixed in commit 5bc7dea**

Corrected assertion message from `want high` to `want max`.

File: packages/wrapper/src/cli.ts [hunk 10]
```diff
@@ -35,11 +48,62 @@ async function main(): Promise<void> {
...
+async function cmdPack(args: string[]): Promise<void> {
+  const sub = args[0];
+  switch (sub) {
+    case 'install': {
+      const isGlobal = args.includes('--global');
+      const force = args.includes('--force');
+      const bnIdx = args.indexOf('--binary-name');
+      const binaryName = bnIdx !== -1 ? args[bnIdx + 1] : undefined;
+      await packInstall({ global: isGlobal, force, binaryName });
+      return;
```
[the remark this answers is not in the document]
Comment (pureliture) [reply packages/wrapper/src/cli.ts:65]: 📋 **Tracked as Issue #35**

Created a follow-up issue for `--binary-name` value validation: https://github.com/pureliture/ai-cli-orch-wrapper/issues/35

This is a P2 enhancement that doesn't block the current PR.

File: CLAUDE.md [hunk 11]
```diff
@@ -4,10 +4,11 @@
...
 
-- **`packages/wrapper/`** — `@aco/wrapper`: Node.js wrapper runtime. Owns the `aco` CLI binary, provider interface (`IProvider`), provider implementations (`GeminiProvider`, `CopilotProvider`), provider registry, and session/task/output lifecycle (`SessionStore`).
-- **`packages/installer/`** — `aco-install`: npm-publishable installer CLI. Provides `aco-install pack install/uninstall/status/setup` and `aco-install provider setup <name>`.
-- **`templates/commands/`** — Source for Claude Code slash command markdown files. Installed to `.claude/commands/` by `aco-install pack install`. Thin wrappers that delegate all logic to `aco run <provider> <command>`.
-- **`templates/prompts/`** — Provider-specific prompt text. Installed to `.claude/aco/prompts/` by `aco-install pack install`.
+- **`packages/wrapper/`** — `@pureliture/ai-cli-orch-wrapper`: Node.js wrapper runtime. Owns the `aco` CLI binary, provider interface (`IProvider`), provider implementations (`GeminiProvider`, `CodexProvider`), provider registry, and session/task/output lifecycle (`SessionStore`). Exposes `aco pack install/setup`, `aco provider setup <name>`, and `aco run ...`.
```
[the remark this answers is not in the document]
Comment (pureliture) [reply CLAUDE.md]: ✅ **Fixed in commit 5bc7dea**

Updated provider list from `GeminiProvider, CodexProvider` to `GeminiProvider, CopilotProvider` to match the actual Node.js registry.

File: docs/contract/go-node-boundary.md [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+### 2. Formatter Routing
+파싱된 frontmatter의 `format` 필드를 기반으로 출력 포맷터를 라우팅한다.
+
+### 3. CLI Flag Validation
+- `--input` 값이 비어있지 않은지 검증
+- `--timeout` 값이 양의 정수인지 검증
+- `--model` 값이 제공된 경우 유효한 모델 식별자인지 검증
+
+### 4. Provider Binary Execution
+프로바이더 바이너리를 `exec.LookPath`로 PATH에서 탐색하고 `syscall.Exec`로 실행한다. 바이너리가 PATH에 없으면 즉시 에러를 반환한다.
```
[the remark this answers is not in the document]
Comment (pureliture) [reply docs/contract/go-node-boundary.md]: ✅ **Fixed in commit 5bc7dea**

Updated the documentation from `syscall.Exec` to accurately describe the `exec.Command/cmd.Start/cmd.Wait` execution model.

**Review by pureliture — COMMENTED** [review]
All Copilot review comments have been addressed:

✅ **Fixed in commit 5bc7dea:**
- Added PATH, HOME to env allowlist for provider CLI compatibility
- Fixed validatePath to check path components, not substrings
- Fixed test message "want high" → "want max"
- Fixed SIGTERM typo
- Updated CLAUDE.md provider list (CodexProvider → CopilotProvider)
- Updated docs/contract/go-node-boundary.md (syscall.Exec → exec.Command)
- Updated callers to use cleaned path from validatePath

📋 **Tracked as Issue #35:**
- --binary-name value validation

**Acknowledged (keeping current behavior):**
- crypto/rand failure returns exit code 1 (intentional)
- SIGTERM test timing (intentional)
- reasoningEffort passthrough (intentional)

All review feedback has been incorporated or documented. This PR is ready for merge.

**Comment by pureliture** [comment]
@codex @gemini 리뷰해

**Comment by pureliture** [comment]
@gemini 리뷰해

File: packages/wrapper/src/util/spawn-stream.ts [hunk 13]
```diff
@@ -44,8 +55,27 @@ export async function* spawnStream(
...
+  // Buffer to accumulate chunks and detect sentinel at the end
+  let buffer = '';
+  let sentinelDetected = false;
+
   for await (const chunk of child.stdout) {
-    yield (chunk as Buffer).toString();
+    const text = (chunk as Buffer).toString();
+    buffer += text;
+    yield text;
+  }
```
[the remark this answers is not in the document]
Comment (pureliture) [reply packages/wrapper/src/util/spawn-stream.ts:72]: ✅ **Fixed in commit 8f4d52c**

Changed from full buffer to last-line-only tracking. Now only the content after the last newline is retained for sentinel detection, eliminating OOM risk on large provider output.

```typescript
let lastLineBuffer = '';
for await (const chunk of child.stdout) {
  const text = (chunk as Buffer).toString();
  lastLineBuffer += text;
  const lastNewlineIndex = lastLineBuffer.lastIndexOf('\n');
  if (lastNewlineIndex !== -1) {
    lastLineBuffer = lastLineBuffer.substring(lastNewlineIndex + 1);
  }
  yield text;
}
```

File: cmd/aco/cmd_delegate.go [hunk 14]
```diff
@@ -249,3 +304,39 @@ func stdinHasData(r io.Reader) bool {
...
+		absPath, err := filepath.Abs(cleaned)
+		if err != nil {
+			return "", fmt.Errorf("invalid file path: %w", err)
+		}
+		absBase, err := filepath.Abs(base)
+		if err != nil {
+			return "", fmt.Errorf("invalid base path: %w", err)
+		}
+		if !strings.HasPrefix(absPath, absBase) {
+			return "", fmt.Errorf("invalid file path: escapes base directory")
```
[the remark this answers is not in the document]
Comment (pureliture) [reply cmd/aco/cmd_delegate.go:343]: ✅ **Fixed in commit 8f4d52c**

Added separator suffix to `absBase` for strict prefix checking. Now `/tmp/foo/` won't match `/tmp/foobar`.

```go
if !strings.HasSuffix(absBase, string(filepath.Separator)) {
    absBase += string(filepath.Separator)
}
```

File: internal/delegate/config.go [hunk 15]
```diff
@@ -225,7 +224,19 @@ func BuildPrompt(spec AgentSpec, input string) (string, error) {
...
+		// Check each path component for ".." to reject actual traversal attempts
+		// while allowing legitimate names like "foo..bar"
+		components := strings.FieldsFunc(seedPath, func(r rune) bool {
+			return r == '/' || r == '\\' || r == rune(filepath.Separator)
+		})
+		for _, component := range components {
+			if component == ".." {
+				return "", fmt.Errorf("invalid promptSeedFile: path traversal not allowed")
+			}
+		}
```
[the remark this answers is not in the document]
Comment (pureliture) [reply internal/delegate/config.go:238]: ✅ **Fixed in commit 8f4d52c**

Added absolute path rejection for `promptSeedFile`. Now rejects paths like `/etc/passwd`.

```go
if filepath.IsAbs(seedPath) {
    return "", fmt.Errorf("promptSeedFile must be relative path")
}
```

File: scripts/verify-contract.ts [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+    if (!reverseEntry) {
+      errors.push(`[DRIFT] TypeScript method "${ts.name}()" has no Go Provider counterpart in GO_TO_TS_MAP`);
+    }
+  }
+
+  return errors;
+}
+
+// ── Main ───────────────────────────────────────────────────────────────────
+
```
[the remark this answers is not in the document]
Comment (pureliture) [reply scripts/verify-contract.ts:155]: ✅ **Fixed in commit 8f4d52c**

Replaced `__dirname` with ESM-compatible approach:

```typescript
import { fileURLToPath } from 'node:url';
const __dirname = path.dirname(fileURLToPath(import.meta.url));
```

File: cmd/aco/cmd_delegate.go [hunk 17]
```diff
@@ -39,18 +42,38 @@ func cmdDelegate(d *deps, args []string) int {
...
 			if i+1 < len(args) {
 				val := args[i+1]
-				// Only reject if it looks like another known flag
-				if val == "--agents-dir" || val == "--formatter" ||
-					val == "--no-formatter" || val == "--no-meta" ||
-					val == "--timeout" {
+				switch val {
+				case "--":
+					// Terminator: no more args
+					i++
```
[the remark this answers is not in the document]
Comment (pureliture) [reply cmd/aco/cmd_delegate.go:56]: **Acknowledged - Current behavior is intentional**

The `--` after `--input` acts as a standard CLI flag terminator. This is a common POSIX convention. If a user needs literal `--` as input, they can use `--input="--"`.

Added a clarifying comment in the code:

```go
case "--":
    // Terminator: stop flag parsing, --input remains empty
    // Note: if user wants literal "--" as input, use --input="--"
    i++
```

File: internal/runner/process.go [hunk 18]
```diff
@@ -27,6 +27,36 @@ import (
...
 
+// envAllowlist defines which environment variables are passed to providers.
+// This prevents sensitive environment variables (API keys, tokens, etc.)
+// from being exposed to provider processes.
+var envAllowlist = map[string]bool{
+	"ACO_TIMEOUT_SECONDS": true,
+	"PATH":                true,
+	"HOME":                true,
+	"USER":                true,
+	"TERM":                true,
```
[the remark this answers is not in the document]
Comment (pureliture) [reply internal/runner/process.go:39]: ✅ **Fixed in commit 8f4d52c**

Added provider auth variables to env allowlist for CI/headless operation:

```go
"GEMINI_API_KEY":   true,
"GITHUB_TOKEN":     true,
"ANTHROPIC_API_KEY": true,
```

File: packages/wrapper/src/util/spawn-stream.ts [hunk 19]
```diff
@@ -44,8 +55,27 @@ export async function* spawnStream(
...
 
+  // Buffer to accumulate chunks and detect sentinel at the end
+  let buffer = '';
+  let sentinelDetected = false;
+
   for await (const chunk of child.stdout) {
-    yield (chunk as Buffer).toString();
+    const text = (chunk as Buffer).toString();
+    buffer += text;
+    yield text;
```
[the remark this answers is not in the document]
Comment (pureliture) [reply packages/wrapper/src/util/spawn-stream.ts:71]: ✅ **Fixed in commit 8f4d52c**

Replaced full stdout buffering with last-line-only tracking. See the sibling Gemini comment for the same fix — both issues addressed in one change.

**Comment by pureliture** [comment]
## Gemini/Codex 리뷰 대응 완료 (commit 8f4d52c)

모든 리뷰 코멘트에 대응하고 resolved 처리했습니다.

### 수정 완료 (5건)

| # | 소스 | 파일 | 이슈 | 조치 |
|---|------|------|------|------|
| 1 | Gemini/Codex | `spawn-stream.ts` | 전체 stdout 버퍼링 OOM 위험 (high/P2) | 마지막 줄만 추적하는 슬라이딩 윈도우로 변경 |
| 2 | Gemini | `cmd_delegate.go:337` | `HasPrefix` 경로 검증 취약점 (security-high) | `absBase`에 separator 추가로 prefix collision 방지 |
| 3 | Gemini | `config.go:237` | 절대 경로 허용 취약점 (security-high) | `filepath.IsAbs`로 절대 경로 차단 |
| 4 | Gemini | `verify-contract.ts:154` | ESM에서 `__dirname` 사용 불가 (medium) | `fileURLToPath(import.meta.url)`로 교체 |
| 5 | Codex | `process.go:38` | provider auth 변수 allowlist 누락 (P1) | `GEMINI_API_KEY`, `GITHUB_TOKEN`, `ANTHROPIC_API_KEY` 추가 |

### 의도적 동작 유지 (1건)

| # | 소스 | 파일 | 이슈 | 조치 |
|---|------|------|------|------|
| 6 | Gemini | `cmd_delegate.go:55` | `--input --` 처리 모호성 (medium) | POSIX 표준 terminator 동작으로 확인, 명확한 주석 추가 |

### 테스트 결과
- ✅ `go test ./...` 전체 통과
- ✅ `npm test` 전체 통과 (38/38)
- ✅ `npm run format:check` 통과

🤖 Generated with [Claude Code](https://claude.com/claude-code)
